keras_cv/layers/preprocessing/random_flip.py

```diff
@@ -0,0 +1,177 @@
+# You may obtain a copy of the License at
+#
+#     https://www.apache.org/licenses/LICENSE-2.0
+#
+# Unless required by applicable law or agreed to in writing, software
+# distributed under the License is distributed on an "AS IS" BASIS,
+# WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+# See the License for the specific language governing permissions and
+# limitations under the License.
+
+import numpy as np
+import tensorflow as tf
+
+from keras_cv import bounding_box
+from keras_cv import utils
+from keras_cv.layers.preprocessing.base_image_augmentation_layer import (
+    BaseImageAugmentationLayer,
+)
+
+# In order to support both unbatched and batched inputs, the horizontal
+# and vertical axis is reverse indexed
+H_AXIS = -3
+W_AXIS = -2
+
+# Defining modes for random flipping
+HORIZONTAL = "horizontal"
+VERTICAL = "vertical"
+HORIZONTAL_AND_VERTICAL = "horizontal_and_vertical"
+
+
+@tf.keras.utils.register_keras_serializable(package="keras_cv")
+class RandomFlip(BaseImageAugmentationLayer):
+    """A preprocessing layer which randomly flips images during training.
+
+    This layer will flip the images horizontally and or vertically based on the
+    `mode` attribute. During inference time, the output will be identical to
+    input. Call the layer with `training=True` to flip the input.
+
+    Input pixel values can be of any range (e.g. `[0., 1.)` or `[0, 255]`) and
+    of integer or floating point dtype. By default, the layer will output
+    floats.
+
+    Input shape:
+      3D (unbatched) or 4D (batched) tensor with shape:
+      `(..., height, width, channels)`, in `"channels_last"` format.
+
+    Output shape:
+      3D (unbatched) or 4D (batched) tensor with shape:
+      `(..., height, width, channels)`, in `"channels_last"` format.
+
+    Arguments:
+      mode: String indicating which flip mode to use. Can be `"horizontal"`,
+        `"vertical"`, or `"horizontal_and_vertical"`. Defaults to
+        `"horizontal_and_vertical"`. `"horizontal"` is a left-right flip and
+        `"vertical"` is a top-bottom flip.
+      seed: Integer. Used to create a random seed.
+    """
+
+    def __init__(
+        self,
+        mode=HORIZONTAL_AND_VERTICAL,
+        seed=None,
+        bounding_box_format=None,
+        **kwargs
+    ):
+        super().__init__(seed=seed, force_generator=True, **kwargs)
+        self.mode = mode
+        if mode == HORIZONTAL:
+            self.horizontal = True
+            self.vertical = False
+        elif mode == VERTICAL:
+            self.horizontal = False
+            self.vertical = True
+        elif mode == HORIZONTAL_AND_VERTICAL:
+            self.horizontal = True
+            self.vertical = True
+        else:
+            raise ValueError(
+                "RandomFlip layer {name} received an unknown mode="
+                "{arg}".format(name=self.name, arg=mode)
+            )
+        self.auto_vectorize = False
```

Comment:
is tf.image.flip_left_right not vectorizable?

Reply:
with auto_vectorize =True, the two tests checking "test_random_flip_vertical_half" and "test_random_flip_horizontal_half" fail. where expected output is way off from actual output.

Reply:
that is really weird... I can't imagine why that would be the case.  This seems to be indicative of a bigger issue under the hood.

Can you please open a GitHub issue to document this, then we can merge?

Reply:
will do!

Reply:
With `expected_output = None` tests are passing also in `auto_vectorize =True`

What are we doing with:

`expected_output[0, :, :, :] = np.flip(input_images[0, :, :, :], axis=0)` ?

Reply:
What is the logic of these `half_flip` test cases?

Reply:
In this case we are flipping just the first image.